src/clj_irods/icat.clj

```diff
@@ -0,0 +1,207 @@
+         (cache/cached-or-agent (:cache irods) #(apply paged-folder-listing* irods user zone path (apply concat more-opts)) (:icat-pool irods)))))
+
+(defn cached-paged-folder-listing
+  [irods user zone path & {:keys [entity-type sort-column sort-direction limit offset info-types] :or {entity-type :any sort-column :base-name sort-direction :desc limit 1000 offset 0 info-types []}}]
+  (let [more-opts {:entity-type entity-type :sort-column sort-column :sort-direction sort-direction :limit limit :offset offset :info-types info-types}]
+    (->> [path ::paged-folder-listing zone user entity-type info-types sort-column sort-direction limit offset]
+         (cache/cached-or-nil (:cache irods)))))
+
+(defn maybe-paged-folder-listing
+  [& args]
+  (delay (deref (apply paged-folder-listing args))))
+
+(defn- resolve*
+  [depth nested-map]
+  (into {} (map
+             (fn [[k v]] [k (if (= depth 0) (if (delay? v) (cache/rethrow-if-error @v) v) (resolve* (dec depth) v))])
+             nested-map)))
+
+(defn all-cached-listings
+  [irods user zone path]
+  (let [all (get-in @(:cache irods) [path ::paged-folder-listing zone user])]
+    (when (seq all)
+    (delay
+      (resolve* 5 all)))))
+
+(defn- filter-listings*
+  ([filters nested-map]
+   (filter-listings* (count filters) filters nested-map))
+  ([depth filters nested-map]
+   (if (zero? depth)
+     nested-map
+     (let [this-filter (nth filters (- (count filters) depth))]
+       (into {}
+             (map (fn [[k v]] [k (filter-listings* (dec depth) filters v)])
+                  (if (nil? this-filter)
+                    nested-map
+                    (filter (fn [[k v]] (= k this-filter)) nested-map))))))))
+
+(defn filtered-cached-listings
+  [irods user zone path & {:keys [entity-type sort-column sort-direction limit offset info-types]}]
+  (let [all (all-cached-listings irods user zone path)]
+    (when all
+      (let [filters [entity-type info-types sort-column sort-direction limit offset]
+            cached  (filter-listings* filters @all)]
+        (when (seq cached)
+          (delay cached))))))
+
+(defn- keys-in [m]
+  (if (map? m)
+    (vec
+      (mapcat (fn [[k v]]
+                (let [sub (keys-in v)
+                      nested (map #(into [k] %) (filter (comp not empty?) sub))]
+                  (if (seq nested)
+                    nested
+                    [[k]])))
+              m))
+    []))
+
+;; NOTE: this merges only directly adjacent ranges -- overlapping ranges will remain separate. This could be less than ideal.
+;; e.g. if we have three ranges, (0 30) (10 40) (20 50), none of them will merge even though we have everything in (0 50)
+;; I'm hoping this doesn't matter much in practice, since it's usually not all that useful to grab overlapping ranges.
+(defn- merge-listing
+  [extant to-merge limit offset]
+  (let [ranges (into {} (mapcat (fn [[k v]] (mapv #(vector (+ % k) [k %]) (keys v))) extant))]
```

Comment:
This makes a map from the ends of ranges (limit + offset) to their path into the map. The map passed into this is only two deep, limit & offset, so it's basically a way to get from "I want a range ending at N" to "there is a range ending at N at a path with limit X and offset Y".